Reduce data redundancy

yotsuya opened the pull request:
This PR modifies some data types to reduce data redundancy. The purpose of this PR is to make it easy to maintain the integrity of data when handling notifications.

## Changes (UPDATE: see https://github.com/iij-ii/direct-hs/pull/62#issuecomment-440606390)

### Implementation

* :new: `data Users`
    - This is a data type for managing list of `User` instead of just `[User]`.
    - This has two fields:
        - `myself`: `User` for the logged-in user
        - `acquaintances`: `[User]` for other users
* `data Client`
    - `clientMe`: removed
        - Instead, `clientUsers` holds `me` (as `myself`)
    - `clientUsers`: data type changed from `[User]` to `Users`
* `data TalkRoom`
    - `talkUsers`: removed
    - :new: `talkUserIds`: `[UserId]` for talk room members
* `data Channel`
    - `channelTalkRoom`: removed

### Interfaces

|                  | Before                    | After                                |
|------------------|---------------------------|--------------------------------------|
| setUsers         | Client -> [User] -> IO () | Client -> Users -> IO ()             |
| getUsers         | Client -> IO [User]       | Client -> IO Users                   |
| getMe            | Client -> IO (Maybe User) | Client -> IO User                    |
| setAcquaintances | -                         | :new: Client -> [User] -> IO ()      |
| getAcquaintances | -                         | :new: Client -> IO [User]            |
| talkUsers        | TalkRoom -> [User]        | :no_entry: removed                          |
| getTalkUsers     | -                         | :new: Client -> TalkRoom -> IO Users |
| channelTalkRoom  | Channel -> TalkRoom       | :no_entry: removed                          |

igrep commented on direct-hs/src/Web/Direct/Client.hs:
```diff
@@ -100,16 +103,29 @@ getTalkRooms :: Client -> IO [TalkRoom]
+getMe :: Client -> IO User
+getMe client = do
+    users <- getUsers client
+    return $ myself users
```
Tips (You don't have to rewrite): In point-free style: `fmap myself . getUsers`

igrep commented on direct-hs-examples/src/ping.hs:
```diff
@@ -39,14 +39,16 @@ handleCreateMessage client (D.Txt "select", _, room, _) = void $ D.sendMessage
-    let ans = T.unlines $ map D.displayName users
+    let
+        ans = T.unlines
+            $ map D.displayName (D.myself users : D.acquaintances users)
```
Perhaps we should have a function like this:

```hs
getUsersList :: Client -> IO [User]
getUsersList client = do
  users <- getUsers client
  return (myself users : acquaintances users)
```

yotsuya replied:
I added helper function `usersList` to construct a user list easily. Does it make sense?

igrep replied:
According to oral discussion:

- Keep `getUsers`' signature `:: Client -> IO [User]` and behavior.
- Save `me` and `acquaintances` in separate `IORef`s.

Rationale: The `Users` type sounds weird. And actually we get `me` and `acquaintances` by separate RPC calls.

yotsuya commented:
The initial design of PR was changed by [the above discussion](https://github.com/iij-ii/direct-hs/pull/62#discussion_r235304731). So, I summarized the changes again.

## Changes

### Implementation

* `data Client`
    - `clientMe`: data type changed from `Maybe User` to `User`
    - `clientUsers`: removed
    - `clientAcquaintances`: :new: `[User]` for acquaintance users (retrieved by `get_acquaintances`)
* `data TalkRoom`
    - `talkUsers`: removed
    - :new: `talkUserIds`: `[UserId]` for talk room members
* `data Channel`
    - `channelTalkRoom`: removed

### Interfaces

|                  | Before                    | After                                |
|------------------|---------------------------|--------------------------------------|
| getMe            | Client -> IO (Maybe User) | Client -> IO User                    |
| setAcquaintances | -                         | :new: Client -> [User] -> IO ()      |
| getAcquaintances | -                         | :new: Client -> IO [User]            |
| setUsers         | Client -> [User] -> IO () | :no_entry: removed                   |
| talkUsers        | TalkRoom -> [User]        | :no_entry: removed                   |
| getTalkUsers     | -                         | :new: Client -> TalkRoom -> IO Users |
| channelTalkRoom  | Channel -> TalkRoom       | :no_entry: removed                   |
